exporter/exporterhelper/internal/unbounded_channel.go

```diff
@@ -0,0 +1,34 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal"
+
+// newUnboundedChannel creates a pair of buffered channels with no cap.
+// Caller should close the input channel when done. That will trigger draining of the output channel.
+// Closing the output channel by client will panic.
+func newUnboundedChannel[T any]() (chan<- T, <-chan T) {
```

Comment:
Why do you need two channels and a goroutine here. I am confused why not just a simple unbounded channel.

Reply:
What kind of simple unbounded channel are you referring to?

Reply:
I had something else in my mind which does not work. What if we create a channel with same Capacity as the "capacity" of the limiter? Every element has at least size 1, so we should be good. Does that mean we are allocating too much memory? Can you check how much memory will allocate?

Reply:
I put a draft with the bounded capacity https://github.com/open-telemetry/opentelemetry-collector/pull/9164 with some benchmarks:

Unbounded:
```
Benchmark_QueueUsage_1M_items_10_250k-10     	       9	 118011278 ns/op	 7497951 B/op	  250017 allocs/op
Benchmark_QueueUsage_1M_items_10_1M-10    	       3	 478627375 ns/op	29998074 B/op	 1000004 allocs/op
Benchmark_QueueUsage_100M_items_10_10M-10    	       1	4688934042 ns/op	299973192 B/op	 9999824 allocs/op
```

Bounded:
```
Benchmark_QueueUsage_1M_items_10_25k-10     	      25	  44122055 ns/op	24003696 B/op	      23 allocs/op
Benchmark_QueueUsage_1M_items_10_1M-10    	       7	 176613071 ns/op	24003396 B/op	      20 allocs/op
Benchmark_QueueUsage_100M_items_10_10M-10    	       1	1693857875 ns/op	2400016832 B/op	      62 allocs/op
```

B/op seems like cumulative. Not sure how to check the max RAM usage. But it is still enough to see that the bounded option takes 2Gb of memory for 100M of the item's capacity configuration value.

The unbounded has a lot of churn due to slice recreation on the overflow. That potentially can be fixed. We could handle the overflow manually and preallocate bigger arrays.

Also, I have an idea of how to integrate the capacity limiter right into the current unbounded channel implementation. It will remove the unnecessary decoupling between capacity limiting and the channel enqueuing. It should also make the code a bit simpler and maybe more performant. I can try that out.

Reply:
For simplicity I would go for the moment with a buffered channel. It has better performance and also seem to have significant less churn (allocations). Let's go with that for the moment.

Reply:
Ok, I updated https://github.com/open-telemetry/opentelemetry-collector/pull/9164 and marked it as ready